src/hotspot/share/utilities/count_trailing_zeros.hpp

```diff
@@ -59,16 +78,32 @@ inline unsigned count_trailing_zeros(uintx x) {
 #pragma intrinsic(_BitScanForward)
 #endif
 
-inline unsigned count_trailing_zeros(uintx x) {
-  assert(x != 0, "precondition");
-  unsigned long index;
+template <typename T> struct CountTrailingZerosImpl<T, 4> {
+  static unsigned doit(T v) {
+    unsigned long index;
+    _BitScanForward(&index, v);
+    return index;
+  }
+};
+
+template <typename T> struct CountTrailingZerosImpl<T, 8> {
+  static unsigned doit(T v) {
+    unsigned long index;
 #ifdef _LP64
-  _BitScanForward64(&index, x);
+    _BitScanForward64(&index, v);
 #else
-  _BitScanForward(&index, x);
+    _BitScanForward(&index, (uint32_t)v);
+    if (index == 0) {
+      _BitScanForward(&index, (uint32_t)(v >> 32));
+      if (index > 0) {
```

Comment:
Shouldn't this check be superfluous if the code previously asserted v != 0, and we know that the lower 32 bits are zero?

Also the MSDN documentation does not state that `index == 0` if no bit has been found, but the return value of the method is 0. I.e. shouldn't the implementation be something like:

```
if (_BitScanForward(&index, x) == 0) { // no bit found? If so, try the upper dword. Otherwise index already contains the result
  _BitScanForward(&index, (uint32_t)(v >> 32));
  // one could assert that the result of this invocation is non-zero, but it should be 1 if v != 0.
  index += 32;
}
return index;
```

Reply:
Yes, I misread the MSDN documentation and it appears index is undefined if there are no bits and we should rely on the return value.

Omitting the cast on `v >> 32` seems fine, but the comment is a bit off since `v` can be a signed type here.

Reply:
You are right about `v` possibly being anything, I did not look correctly. But then the new code does indeed need some cast here because the result of `something signed >> 32` is compiler dependent (at least until C++20).

Since this code path is specific to Visual C++, it might still be true that this is the case for all platforms (ie. even ARM64), but I'd prefer the explicit cast.

Reply:
Right, I think what I have works on Windows x64, but let's be explicit